### ISSUE TITLE
can mpi-operator support horovodrun?

### ISSUE THREAD

**Issue by davidstack** [kubeflow/mpi-operator#327]
mpi-operator support mpirun command, will mpi-operator support howovodrun ? such as

`
horovodrun -np 16 -H server1:4,server2:4,server3:4,server4:4 python train.py
`

**Comment by terrytangyuan** [comment]
If it's fully compatible with mpirun then it should work. Perhaps give it a try?

**Comment by davidstack** [comment]
yes, i have got a test. if i use this command, the job will run successfully. but i do not know whether other horovod functions can work correctly. And we should know the hostname of the worker.
`
horovodrun  -np 2 -H mpi-d3ec808b-worker-0:1,mpi-d3ec808b-worker-1:1  python /horovod/tensorflow_mnist.py
`

**Comment by asahalyft** [comment]
@davidstack I started out with this example - it should be able to run a mpi horovod job on your K8s cluster provides you alreadt have the mpi-operator applied on the K8s cluster.
```
# ------- Create GPU Distributed MPI Job ------ #
apiVersion: kubeflow.org/v1alpha2
kind: MPIJob
metadata: 
  name: tf2-keras-mnist-mpi-gpu
spec: 
  slotsPerWorker: 1
  cleanPodPolicy: Running
  mpiReplicaSpecs: 
    Launcher: 
      replicas: 1
      template: 
        spec: 
          containers: 
            - name: keras-mnist-mpi-launcher
              image: "docker.io/horovod/horovod:0.19.3-tf2.1.0-torch-mxnet1.6.0-py3.6-gpu"
              command: 
                - mpirun
              args: 
                - "-np"
                - "2"
                - "--allow-run-as-root"
                - "-bind-to"
                - none
                - "-map-by"
                - slot
                - "-x"
                - NCCL_DEBUG=INFO
                - "-x"
                - LD_LIBRARY_PATH
                - "-x"
                - PATH
                - "-x"
                - NCCL_SOCKET_IFNAME=eth0
                - "-mca"
                - pml
                - ob1
                - "-mca"
                - btl
                - ^openib
                - python
                - /examples/tensorflow2_keras_mnist.py
              resources: 
                limits: 
                  cpu: 1
                  memory: 2Gi
    Worker: 
      replicas: 2
      template: 
        spec: 
          containers: 
            - name: keras-mnist-mpi-worker
              image: "docker.io/horovod/horovod:0.19.3-tf2.1.0-torch-mxnet1.6.0-py3.6-gpu"
              resources: 
                limits: 
                  nvidia.com/gpu: 1
          tolerations: 
            - key: lyft.net/gpu
              operator: Equal
              value: dedicated
              effect: NoSchedule
```

You may have to change the `NCCL_SOCKET_IFNAME=eth0` based on your setup.

**Comment by davidstack** [comment]
@asahalyft  yes,you are right,but i want to only use horovodrun command

**Comment by zw0610** [comment]
> yes, i have got a test. if i use this command, the job will run successfully. but i do not know whether other horovod functions can work correctly. And we should know the hostname of the worker.
> `horovodrun -np 2 -H mpi-d3ec808b-worker-0:1,mpi-d3ec808b-worker-1:1 python /horovod/tensorflow_mnist.py`

@davidstack `horovodrun` and mpi-operator are compatible now. I took a further test:

```shell
root@tensorflow-mnist-launcher-kw4jr:/examples# horovodrun -np 2 -hostfile /etc/mpi/hostfile hostname
+ POD_NAME=tensorflow-mnist-worker-0
+ [ t = - ]
+ shift
+ /opt/kube/kubectl cp /opt/kube/hosts tensorflow-mnist-worker-0:/etc/hosts_of_nodes
+ POD_NAME=tensorflow-mnist-worker-1
+ [ t = - ]
+ shift
+ /opt/kube/kubectl cp /opt/kube/hosts tensorflow-mnist-worker-1:/etc/hosts_of_nodes
+ /opt/kube/kubectl exec tensorflow-mnist-worker-0 -- /bin/sh -c cat /etc/hosts_of_nodes >> /etc/hosts &&     PATH=/usr/local/bin:$PATH ; export PATH ; LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH ; export LD_LIBRARY_PATH ; DYLD_LIBRARY_PATH=/usr/local/lib:$DYLD_LIBRARY_PATH ; export DYLD_LIBRARY_PATH ;   /usr/local/bin/orted -mca ess "env" -mca ess_base_jobid "124190720" -mca ess_base_vpid 1 -mca ess_base_num_procs "3" -mca orte_node_regex "tensorflow-mnist-launcher-kw[1:4]jr,tensorflow-mnist-worker-[1:0-1]@0(3)" -mca orte_hnp_uri "124190720.0;tcp://10.244.2.50:51787" -mca pml "ob1" -mca btl "^openib" -mca btl_tcp_if_include "eth0" -mca plm "rsh" --tree-spawn -mca orte_parent_uri "124190720.0;tcp://10.244.2.50:51787" -mca orte_default_hostfile "/etc/mpi/hostfile" -mca plm_rsh_agent "/etc/mpi/kubexec.sh" -mca orte_tag_output "1" -mca hwloc_base_binding_policy "none" -mca rmaps_base_mapping_policy "slot" -mca pmix "^s1,s2,cray,isolated"
+ /opt/kube/kubectl exec tensorflow-mnist-worker-1 -- /bin/sh -c cat /etc/hosts_of_nodes >> /etc/hosts &&     PATH=/usr/local/bin:$PATH ; export PATH ; LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH ; export LD_LIBRARY_PATH ; DYLD_LIBRARY_PATH=/usr/local/lib:$DYLD_LIBRARY_PATH ; export DYLD_LIBRARY_PATH ;   /usr/local/bin/orted -mca ess "env" -mca ess_base_jobid "124190720" -mca ess_base_vpid 2 -mca ess_base_num_procs "3" -mca orte_node_regex "tensorflow-mnist-launcher-kw[1:4]jr,tensorflow-mnist-worker-[1:0-1]@0(3)" -mca orte_hnp_uri "124190720.0;tcp://10.244.2.50:51787" -mca pml "ob1" -mca btl "^openib" -mca btl_tcp_if_include "eth0" -mca plm "rsh" --tree-spawn -mca orte_parent_uri "124190720.0;tcp://10.244.2.50:51787" -mca orte_default_hostfile "/etc/mpi/hostfile" -mca plm_rsh_agent "/etc/mpi/kubexec.sh" -mca orte_tag_output "1" -mca hwloc_base_binding_policy "none" -mca rmaps_base_mapping_policy "slot" -mca pmix "^s1,s2,cray,isolated"
[1,0]<stdout>:tensorflow-mnist-worker-0
[1,1]<stdout>:tensorflow-mnist-worker-1
```

Please keep in mind that option `-hostfile /etc/mpi/hostfile` is a must.

**Comment by terrytangyuan** [comment]
@zw0610 Would you like to add an example illustrating how to use `horovodrun` with MPI operator?

**Comment by zw0610** [comment]
> @zw0610 Would you like to add an example illustrating how to use `horovodrun` with MPI operator?

Sure. Let me compose another example and merge it into `examples/horovod`.

**Comment by davidstack** [comment]
@zw0610 i have a test, but it failed.
`
root@mpi-feea39f9-launcher:/examples# horovodrun -np 2 -hostfile=/etc/mpi/hostfile hostname                                            
ssh not successful for host mpi-feea39f9-worker-1:
ssh: Could not resolve hostname mpi-feea39f9-worker-1: Name or service not known

ssh not successful for host mpi-feea39f9-worker-0:
ssh: Could not resolve hostname mpi-feea39f9-worker-0: Name or service not known

`

the env in this lanucher is 

`
HOSTNAME=mpi-feea39f9-launcher
OMPI_MCA_orte_default_hostfile=/etc/mpi/hostfile
NVIDIA_VISIBLE_DEVICES=none
KUBERNETES_PORT_443_TCP_PROTO=tcp
KUBERNETES_PORT_443_TCP_ADDR=10.233.0.1
NCCL_VERSION=2.4.8
OMPI_MCA_plm_rsh_agent=/etc/mpi/kubexec.sh
KUBERNETES_PORT=tcp://10.233.0.1:443
PWD=/examples
HOME=/root
TENSORFLOW_VERSION=1.14.0
KUBERNETES_SERVICE_PORT_HTTPS=443
KUBERNETES_PORT_443_TCP_PORT=443
LIBRARY_PATH=/usr/local/cuda/lib64/stubs
KUBERNETES_PORT_443_TCP=tcp://10.233.0.1:443
TERM=xterm
CUDA_PKG_VERSION=10-1=10.1.243-1
CUDA_VERSION=10.1.243
NVIDIA_DRIVER_CAPABILITIES=compute,utility
PYTHON_VERSION=3.6
SHLVL=1
NVIDIA_REQUIRE_CUDA=cuda>=10.1 brand=tesla,driver>=384,driver<385 brand=tesla,driver>=396,driver<397 brand=tesla,driver>=410,driver<411
KUBERNETES_SERVICE_PORT=443
PATH=/usr/local/nvidia/bin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
KUBERNETES_SERVICE_HOST=10.233.0.1
`

the pip list in this container is

`
Package              Version 
-------------------- --------
absl-py              0.8.1   
astor                0.8.0   
attrs                19.3.0  
backcall             0.1.0   
bleach               3.1.0   
cffi                 1.13.1  
cloudpickle          1.2.2   
cycler               0.10.0  
decorator            4.4.0   
defusedxml           0.6.0   
entrypoints          0.3     
future               0.18.1  
gast                 0.3.2   
google-pasta         0.1.7   
grpcio               1.24.1  
h5py                 2.10.0  
horovod              0.18.1  
importlib-metadata   0.23    
ipykernel            5.1.3   
ipython              7.8.0   
ipython-genutils     0.2.0   
ipywidgets           7.5.1   
jedi                 0.15.1  
Jinja2               2.10.3  
json5                0.8.5   
jsonschema           3.1.1   
jupyter              1.0.0   
jupyter-client       5.3.4   
jupyter-console      6.0.0   
jupyter-core         4.6.0   
jupyterlab           1.1.4   
jupyterlab-server    1.0.6   
Keras                2.3.1   
Keras-Applications   1.0.8   
Keras-Preprocessing  1.1.0   
kiwisolver           1.1.0   
Markdown             3.1.1   
MarkupSafe           1.1.1   
matplotlib           3.1.1   
mistune              0.8.4   
more-itertools       7.2.0   
mpi4py               3.0.2   
mxboard              0.1.0   
nbconvert            5.6.0   
nbformat             4.4.0   
notebook             6.0.1   
numpy                1.17.3  
opencv-python        4.1.1.26
pandocfilters        1.4.2   
parso                0.5.1   
pexpect              4.7.0   
pickleshare          0.7.5   
Pillow               6.2.1   
pip                  19.3.1  
prometheus-client    0.7.1   
prompt-toolkit       2.0.10  
protobuf             3.10.0  
psutil               5.6.3   
ptyprocess           0.6.0   
pycparser            2.19    
Pygments             2.4.2   
pygobject            3.26.1  
pyparsing            2.4.2   
pyrsistent           0.15.4  
python-apt           1.6.4   
python-dateutil      2.8.0   
PyYAML               5.1.2   
pyzmq                18.1.0  
qtconsole            4.5.5   
scipy                1.3.1   
Send2Trash           1.5.0   
setuptools           41.4.0  
six                  1.12.0  
tensorboard          1.14.0  
tensorboardX         1.9     
tensorflow-estimator 1.14.0  
tensorflow-gpu       1.14.0  
termcolor            1.1.0   
terminado            0.8.2   
testpath             0.4.2   
tornado              6.0.3   
tqdm                 4.36.1  
traitlets            4.3.3   
typing               3.7.4.1 
wcwidth              0.1.7   
webencodings         0.5.1   
Werkzeug             0.16.0  
wheel                0.33.6  
widgetsnbextension   3.5.1   
wrapt                1.11.2  
zipp                 0.6.0

`

how can i use the /etc/mpi/kubexec.sh  to ssh to these workers? thanks

**Comment by gaocegege** [comment]
@davidstack  

Please try Horovod v0.20.0 or above

We have a PR in Horvod to support kubeexec in Kubeflow: https://github.com/horovod/horovod/pull/2199

And it is released in v0.20.0

**Comment by davidstack** [comment]
@gaocegege got it,thanks
